File: lib/dashboard/heatmap-range.test.ts
import { describe, expect, it } from 'vitest';
import { chooseRenderedHeatmapRange } from './heatmap-range';

const WEEK_COUNT = { '30d': 5, '90d': 14, year: 53 };

describe('chooseRenderedHeatmapRange', () => {
  it('keeps the full year when the viewport can fit it', () => {
    expect(
      chooseRenderedHeatmapRange({
        preferredRange: 'year',
        availableWidth: 900,
        weekCount: WEEK_COUNT,
      })
    ).toBe('year');
  });

  it('keeps the full year when width fits even if height is compact', () => {
    expect(
      chooseRenderedHeatmapRange({
        preferredRange: 'year',
        availableWidth: 900,
        availableHeight: 180,
        weekCount: WEEK_COUNT,
      })
    ).toBe('year');
  });

  it('falls back from year to 90 days when the year is too dense', () => {
    expect(
      chooseRenderedHeatmapRange({
        preferredRange: 'year',
        availableWidth: 360,
        weekCount: WEEK_COUNT,
      })
    ).toBe('90d');
  });

  it('keeps explicit 30 day preference even on wide screens', () => {
    expect(
      chooseRenderedHeatmapRange({
        preferredRange: '30d',
        availableWidth: 900,
        weekCount: WEEK_COUNT,
      })
    ).toBe('30d');
  });
});

File: apps/mobile-flutter/test/features/dashboard/heatmap_range_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:kallo_mobile/features/dashboard/logic/heatmap_range.dart';
import 'package:kallo_mobile/models/profile/heatmap.dart';

/// Content width the card gives the grid at a given device width: screen less
/// the 12pt screen padding, the 16pt card padding, the day-label gutter and its
/// 4pt spacer — the same arithmetic `AdherenceHeatmap` does.
double _content(double screenWidth) => screenWidth - 12 * 2 - 16 * 2 - 21 - 4;

void main() {
  HeatmapRange choose(double screenWidth) => chooseRenderedHeatmapRange(
    preferredRange: HeatmapRange.year,
    availableWidth: _content(screenWidth),
  );

  group('chooseRenderedHeatmapRange', () {
    test('every phone width stays on the 90-day window', () {
      // 320 (SE) through 440 (16 Pro Max). A year at 53 columns cannot reach
      // its 10px floor on any of them, so phone rendering is untouched.
      for (final width in [320.0, 360.0, 375.0, 390.0, 402.0, 430.0, 440.0]) {
        expect(choose(width), HeatmapRange.d90, reason: '$width pt');
      }
    });

    test('tablet widths buy history instead of bigger cells', () {
      for (final width in [744.0, 834.0, 1024.0, 1366.0]) {
        expect(choose(width), HeatmapRange.year, reason: '$width pt');
      }
    });

    test('falls back to 30d when even 90d cannot read', () {
      expect(
        chooseRenderedHeatmapRange(
          preferredRange: HeatmapRange.year,
          availableWidth: 120,
        ),
        HeatmapRange.d30,
      );
    });

    test('a 30d preference is never widened', () {
      expect(
        chooseRenderedHeatmapRange(
          preferredRange: HeatmapRange.d30,
          availableWidth: 4000,
        ),
        HeatmapRange.d30,
      );
    });

    test('the ceiling keeps a tablet cell from inflating', () {
      // 90d at iPad width would be ~51px without a cap; the cap is what makes
      // the wider range the better answer rather than a taller card.
      expect(heatmapMaxCell[HeatmapRange.d90], lessThan(51));
    });
  });
}
